### PR TITLE
Fix rebalancer failure counter in async scenario

### LINKED ISSUE #2062: Waged rebalanceFailureCount is not increased when global rebalance failed
### Describe the bug
When global rebalance failed, Waged metrics rebalanceFailureCount is not increased. This failure is not tacked by metrics.

### PR DESCRIPTION
**Pull request by NealSun96**
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Fixes #2062 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Async processes cannot propagate exceptions upwards, so we need to explicitly increment the failure counters during exception handling in the submission block. Since partial rebalance is turning into async, this PR fixes both the old problem (global rebalance missing rebalance failure counter) and the new problem. 

### Tests

- The following is the result of the "mvn test" command on the appropriate module:

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java [hunk 1]
```diff
@@ -460,6 +460,7 @@ private void globalRebalance(ResourceControllerDataProvider clusterData, Map<Str
           doGlobalRebalance(clusterData, resourceMap, algorithm, currentStateOutput, !waitForGlobalRebalance,
               clusterChanges);
         } catch (HelixRebalanceException e) {
+          _rebalanceFailureCount.increment(1L);
```
Comment (qqu0127) [inline helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java:463]: For my own understanding, how was it working in the past? Did we simply skip the error and not counting?

Comment (NealSun96) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedRebalancer.java:463]: Yes, Global Rebalance failure was never accounted for in the failure metric. See #2062 .

**Review by qqu0127 — APPROVED** [review]
LGTM, also shall we consider cherry-pick this to the master branch?

**Comment by NealSun96** [comment]
@qqu0127 This will be merged into master with feature branch, which happens now. :)

**Comment by jiajunwang** [comment]
If both baseline and partial rebalance fail, will we count 2 times instead
of 1?

Best Regards,
Jiajun

**Comment by NealSun96** [comment]
@jiajunwang missed your message: yes we will count 2 times. Conceptually, given that these pipelines are separate async pipelines now, counting 2 times does make sense (rebalance failed on 2 occasions with possibly different settings). Operationally, since the metric is used to indicate errors, counting 2 times is mostly similar to counting 1 time (and definitely better than counting 0 times).

**Comment by jiajunwang** [comment]
Agreed. Not a concern.

**Pull request merged**